src/libvirt-dbus.js

```diff
@@ -1568,8 +1570,32 @@ export function storageVolumeDelete(connectionName, poolName, volName) {
 }
 
 export function vmInterfaceAddresses(connectionName, objPath) {
-    // 'lease' source will work only for those domains which take IP address from libvirt spawned dnsmasq (i.e. have <interface type='network'/>)
-    return call(connectionName, objPath, 'org.libvirt.Domain', 'InterfaceAddresses', [Enum.VIR_DOMAIN_INTERFACE_ADDRESSES_SRC_LEASE, 0], { timeout, type: 'uu' });
+    const error = [];
+    return call(connectionName, objPath, 'org.libvirt.Domain', 'InterfaceAddresses', [Enum.VIR_DOMAIN_INTERFACE_ADDRESSES_SRC_LEASE, 0], { timeout, type: 'uu' })
+            .then(res => {
+                if (!res || res[0].length == 0)
+                    return Promise.reject(new Error("Failed to get interface addresses from lease source"));
+
+                return Promise.resolve(res);
+            })
+            .catch(exc => {
+                error.push(exc.message);
+                return call(connectionName, objPath, 'org.libvirt.Domain', 'InterfaceAddresses', [Enum.VIR_DOMAIN_INTERFACE_ADDRESSES_SRC_ARP, 0], { timeout, type: 'uu' });
+            })
+            .then(res => {
+                if (!res || res[0].length == 0)
+                    return Promise.reject(new Error("Failed to get interface addresses from arp source"));
+
+                return Promise.resolve(res);
+            })
+            .catch(exc => {
+                error.push(exc.message);
+                return call(connectionName, objPath, 'org.libvirt.Domain', 'InterfaceAddresses', [Enum.VIR_DOMAIN_INTERFACE_ADDRESSES_SRC_AGENT, 0], { timeout, type: 'uu' });
+            })
+            .catch(exc => {
+                error.push(exc.message);
+                return Promise.reject(new Error(error.join(", ")));
+            });
```

Comment:
Hmm I can also use Promise.allSettled() instead. That will make it slightly simpler I feel.

Reply:
Followed this approach indeed